File: hypo71py/core/fortran/tinorm.f
      FUNCTION TINORM(ALPHA)                                            TINORM
      DIMENSION A(3),B(3)                                               TINORM
      DATA(A(I),I=1,3)/.010328,.802853,2.515517/,(B(I),I=1,3)/.0010308, TINORM
     1.189269,1.432788/                                                 TINORM
C-----------------------------------------------------------------------TINORM
C     APPROXIMATION TO INVERSE NORMAL DISTRIBUTION                      TINORM
C-----------------------------------------------------------------------TINORM
      IF(.NOT.(ALPHA.GT.0..AND.ALPHA.LT.1.)) GO TO 1                    TINORM
      X=ALPHA                                                           TINORM
      IF(X.GT..5) X=1.-X                                                TINORM
      X=SQRT(-2.*ALOG(X))                                               TINORM
      TINORM=X-(A(3)+X*(A(2)+X*A(1)))/(1.+X*(B(3)+X*(B(2)+X* B(1))))    TINORM
c     CALL OVERFL(I)                                                    TINORM
C          CHANGE     R.VERBEIREN   23/02/83
C                                   11/04/90 - microvax 
C     IF(I.EQ.1) RETURN 2                                               TINORM
c     IF(I.EQ.1) RETURN 1
      IF(ALPHA.LT..5) TINORM=-TINORM                                    TINORM
    1 RETURN                                                            TINORM
C   1 RETURN 2                                                          TINORM
c   1 RETURN 1
      END                                                               TINORM

File: hypo71py/core/tinorm.py
"""
"""

import numpy as np


def TINORM(ALPHA):
	"""
	APPROXIMATION TO INVERSE CDF OF STANDARD NORMAL DISTRIBUTION
	"""
	A = np.array([.010328, .802853, 2.515517])
	B = np.array([.0010308, .189269, 1.432788])

	ALPHA = np.asarray(ALPHA)

	assert ((ALPHA > 0) & (ALPHA < 1)).all()

	X = ALPHA.copy()
	over_half_idxs = ALPHA > .5
	X[over_half_idxs] = 1 - X[over_half_idxs]
	X = np.sqrt(-2 * np.log(X))

	TINORM = X - (A[2]+X*(A[1]+X*A[0])) / (1.+X*(B[2]+X*(B[1]+X* B[0])))
	TINORM = np.asarray(TINORM)
	below_half_idxs = ALPHA < .5
	TINORM[below_half_idxs] = -TINORM[below_half_idxs]

	return TINORM



if __name__ == "__main__":
	import pylab

	alpha1 = np.linspace(0, 1, 1000)[1:-1]
	tinorm1 = TINORM(alpha1)
	pylab.plot(alpha1, tinorm1)
	alpha2 = np.random.sample(1000)
	tinorm2 = TINORM(alpha2)
	pylab.plot(alpha2, tinorm2, 'ro')
	pylab.show()
